Create new Atoms in get_structure if size changes

pmrv opened the pull request:
pbc and species taken from job.structure, because they're not saved in
the HDF5.

This enables to use `get_structure` on an interactive jobs, when you put in differently sized structures.  It does not help sadly with the problem of needing all species present in the initial structure.

pmrv committed:
Create new Atoms in get_structure if size changes

pbc and species taken from job.structure, because they're not saved in
the HDF5.

sudarsan-surendralal commented:
> 
> 
> pbc and species taken from job.structure, because they're not saved in
> the HDF5.
> 
> This enables to use `get_structure` on an interactive jobs, when you put in differently sized structures. It does not help sadly with the problem of needing all species present in the initial structure.

I guess we're talking about a grand canonical situation where the species and the count of each species changes? Shouldn't there be something like `self.output.species` for such jubs then? Wouldn't this solve the problem that all species should be present at first?

pmrv commented:
In this case it's more about calculating energies for arbitrary structures via an interactive job, but grand-canonical would be another application.

I had wanted to use `output.species` but in my quick test it was not populated, i.e. it was an empty list.  Maybe I missed something?

liamhuber reviewed (approved):
Not a corner of the code base I'm super familiar with, but it looks good to me.

sudarsan-surendralal commented on pyiron_atomistics/atomistics/job/atomistic.py:
```diff
@@ -612,19 +612,23 @@ def get_final_structure(self):
+            snapshot = Atoms(species=self.structure.species, indices=indices,
+                             positions=np.zeros(indices.shape + (3,)), cell=cell, pbc=self.structure.pbc)
+        else:
+            snapshot = self.structure.copy()
```
Shouldn't you set the cell of the snapshot here? 
```suggestion
            snapshot = self.structure.copy()
            if cell is not None:
                snapshot.cell = cell
```

pmrv replied:
Jup, added this and also the indices.

sudarsan-surendralal commented:
> I had wanted to use `output.species` but in my quick test it was not populated, i.e. it was an empty list. Maybe I missed something?

I realize that the interactive codes (for now LAMMPS and SPHINX) don't store the changing species info in the output (which is bad IMO). Would it be difficult to implement this (perhaps even in a separate PR)?

pmrv committed:
Set cell/indices/species if structure is copied from self.structure

pmrv commented:
> > I had wanted to use `output.species` but in my quick test it was not populated, i.e. it was an empty list. Maybe I missed something?
> 
> I realize that the interactive codes (for now LAMMPS and SPHINX) don't store the changing species info in the output (which is bad IMO). Would it be difficult to implement this (perhaps even in a separate PR)?

I've just checked that this is also the case for non-interactive run jobs, e.g. Lammps & calc_minimize.  I've used `job['output/structure/species']` for now which should be the same as `job.structure.species`, but I agree that this should be addressed in another PR.

sudarsan-surendralal commented on pyiron_atomistics/atomistics/job/atomistic.py:
```diff
@@ -612,19 +612,30 @@ def get_final_structure(self):
+            if indices is not None:
+                snapshot.indices = indices
+            if species is not None:
+                snapshot.species = species
```
You can't set the species directly and have to use `set_species(species)` (This should fix the tests)

sudarsan-surendralal replied:
Sorry, I realize that this doesn't work too since the species list should be a list of `ChemicalElement` instances and not symbols. IMO since this line basically does nothing since we use the original species info, it's probably best to remove this line.

pmrv committed:
Properly set species

pmrv committed:
Three time's a charm

sudarsan-surendralal commented:
Everything seems to work fine now! But still, I wonder if it's possible to add some tests where we check get structure returns the exact thing we (cell, species, and positions). For instance , we fail to capture the fact that we didn't update the cell (based on the initial commits where the tests passed). Sorry for insisting on this but this is a function that pretty much every pyiron workflow uses!

pmrv committed:
Add get_structure test

Since there have to be some output structures for the function to work,
I've added a packed project that contains an interactive lammps job on
Fe di- & trimers each with different cell sizes.

pmrv commented:
I've added a simple test case, but it required me to add a packed project and some tango around how `project.unpack` works right now.

sudarsan-surendralal commented on tests/atomistics/job/test_atomistic.py:
```diff
@@ -69,5 +69,26 @@ def test_get_displacements(self):
+        # have to do extra tango because Project.unpack is weird right now
+        cwd = os.curdir
+        to_dir = os.path.join(self.project.path, "..")
+        shutil.copy("tests/static/lammps_test_files/get_structure_test.tar.gz", to_dir)
```
You haven't imported `shutil` in this file!

pmrv replied:
Ah, the curse of copy-paste, fixed.

pmrv committed:
Fix import

sudarsan-surendralal reviewed (approved):
Looks great now!

niklassiemer commented:
Depending on the start directory of the tests, one may get a `FileNotFoundError` due to "tests/static/lammps_test_files/get_structure_test.tar.gz".
